src/System.Linq/src/System/Linq/Union.cs

```diff
@@ -59,68 +83,52 @@ private bool GetNext()
                 return false;
             }
 
-            public override bool MoveNext()
+            public sealed override bool MoveNext()
             {
-                switch(state)
+                if (state == 1)
                 {
-                    case 1:
-                        _enumerator = _first.GetEnumerator();
-                        if (_enumerator.MoveNext())
-                        {
-                            state = 2;
-                        }
-                        else
+                    for (IEnumerable<TSource> enumerable = GetEnumerable(0); enumerable != null; enumerable = GetEnumerable(state - 1))
+                    {
+                        IEnumerator<TSource> enumerator = enumerable.GetEnumerator();
+                        ++state;
```

Comment:
Maybe `UnionIteratorN.Union()` should have:

    if (_nextIndex > int.MaxValue - 2) return new UnionIterator2<TSource>(this, next);

And the same for `Concat` (which otherwise perhaps should be `checked(_nextIndex + 2 - 1)` instead of `checked(_nextIndex + 1)` to throw one earlier and not overflow in the enumerating, unless my mental arithmetic is off).

If we're being careful about what happens after two billion calls, we should probably consider throwing an undesirable observable change to what was there two days ago.